semantic/src/main/kotlin/tools/samt/semantic/SemanticModel.kt

```diff
@@ -0,0 +1,597 @@
+                                    }
+
+                                    "size" -> {
+                                        if (constraint.arguments.size != 1 || constraint.arguments.firstOrNull() !is RangeExpressionNode) {
+                                            controller.createContext(expression.location.source).error {
+                                                message("Size constraint must have exactly one size argument")
+                                                highlight("invalid constraint", constraint.location)
+                                                help("A valid constraint would be size(1..10)")
+                                            }
+                                            return@mapNotNull null
+                                        }
+                                        return@mapNotNull createSize(constraint.arguments.first() as RangeExpressionNode)
+                                    }
+
+                                    "pattern" -> {
+                                        if (constraint.arguments.size != 1 || constraint.arguments.firstOrNull() !is StringNode) {
+                                            controller.createContext(expression.location.source).error {
+                                                message("Pattern constraint must have exactly one string argument")
+                                                highlight("invalid constraint", constraint.location)
+                                                help("A valid constraint would be pattern(\"a-z\")")
+                                            }
+                                            return@mapNotNull null
+                                        }
+                                        return@mapNotNull createPattern(constraint.arguments.first() as StringNode)
+                                    }
+
+                                    "value" -> {
+                                        if (constraint.arguments.size != 1) {
+                                            controller.createContext(expression.location.source).error {
+                                                message("value constraint must have exactly one argument")
+                                                highlight("invalid constraint", constraint.location)
+                                            }
+                                            return@mapNotNull null
+                                        }
+                                        return@mapNotNull createValue(constraint.arguments.first())
+                                    }
+
+                                    is String -> {
+                                        controller.createContext(expression.location.source).error {
+                                            message("Constraint with name '${name}' does not exist")
+                                            highlight("unknown constraint", constraint.base.location)
```

Comment:
Add a `help` annotation showing all available constraint types that are applicable here.

Reply:
help was added